keps/sig-scheduling/5055-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,987 @@
+    Key string
+
+    // The taint value corresponding to the taint key.
+    // Must be a label value.
+    //
+    // +optional
+    Value string
+
+    // The effect of the taint on claims and pods that do not tolerate the taint.
+    // Valid effects are NoSchedule and NoExecute. PreferNoSchedule as used for
+    // nodes is not valid here.
+    //
+    // +required
+    Effect TaintEffect
+
+    // ^^^^
+    //
+    // Implementing PreferNoSchedule would depend on a scoring solution for DRA.
+    // It might get added as part of that.
+
+    // TimeAdded represents the time at which the taint was added.
+    // It is only written for NoExecute taints.
+    //
+    // +optional
+    TimeAdded *metav1.Time
+}
+```
+
+Taint has the exact same fields as a v1.Taint, but the description is a bit
+different. In particular, PreferNoSchedule is not valid.
+
+As with node taints, the key is typically a short string. The meaning of the value
+depends on the key. It is allowed to be longer. The core v1 API does not impose
+length limitations for these fields. The `resource.k8s.io` API does.
+
+Tolerations get added to a DeviceRequest:
+
+```Go
+type DeviceRequest struct {
+    ...
+
+    // If specified, the request's tolerations.
+    //
+    // Tolerations for NoSchedule are required to allocate a
+    // device which has a taint with that effect. The same applies
+    // to NoExecute.
+    //
+    // In addition, should any of the allocated devices get tainted
+    // with NoExecute after allocation and that effect is not tolerated,
+    // then all pods consuming the ResourceClaim get deleted to evict
+    // them. The scheduler will not let new pods reserve the claim while
+    // it has these tainted devices. Once all pods are evicted, the
+    // claim will get deallocated.
+    //
+	// +optional
+	// +listType=atomic
+	Tolerations []Toleration `json:"tolerations,omitempty" protobuf:"bytes,22,opt,name=tolerations"`
+}
+
+// The ResourceClaim this Toleration is attached to tolerate any taint that matches
+// the triple <key,value,effect> using the matching operator <operator>.
+type Toleration struct {
+	// Key is the taint key that the toleration applies to. Empty means match all taint keys.
+	// If the key is empty, operator must be Exists; this combination means to match all values and all keys.
+    // Must be a label name.
+    //
+	// +optional
+	Key string
+
+	// Operator represents a key's relationship to the value.
+	// Valid operators are Exists and Equal. Defaults to Equal.
+	// Exists is equivalent to wildcard for value, so that a ResourceClaim can
+	// tolerate all taints of a particular category.
+    //
+	// +optional
+	Operator TolerationOperator
+
+	// Value is the taint value the toleration matches to.
+	// If the operator is Exists, the value should be empty, otherwise just a regular string.
+    // Must be a label value.
+    //
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,3,opt,name=value"`
```

Comment:
What kind of values are expected here?  E.g. `true`  / `false` are longer than 3 bytes...

Reply:
`3` is the protobuf index number, not the maximum length. Which reminds me, I should remove the struct tags because they don't add value here in the KEP...

The value is whatever the creator of a taint sets as value. The API itself doesn't care as long as it has the form of a Kubernetes label value. What that is exactly isn't defined for node taints ever, so presumably other documentation covers that or its considered common knowledge. It's the same as for labels.

Reply:
> Which reminds me, I should remove the struct tags because they don't add value here in the KEP...

Fixed, together with tab vs spaces indention.